### PR TITLE
[WIP]Added official guestbook example

### PR DESCRIPTION
**Pull request by surajnarwade**
* guestbook-official is referred from https://kubernetes.io/docs/tutorials/stateless-application/guestbook/
  This will work with Kubernetes only

* guestbook-go is referred from https://github.com/kubernetes/kubernetes/tree/master/examples/guestbook-go
  This will work with both Kubernetes as well as OpenShift

### CODE REVIEW

File: examples/guestbook-go/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (cdrage) [inline examples/guestbook-go/README.md]: please remove all of these newlines ^^

File: examples/guestbook-go/redis-master.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+containers:
+- name: redis-master
+  image: redis:2.8.23
+  
+
+services:
+- name: redis-master
+  portMappings: 
+  - 6379:6379
+
```
Comment (cdrage) [inline examples/guestbook-go/redis-master.yml]: remove this newline

File: examples/guestbook-go/redis-master.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+name: redis-master
+
+labels:
+  app: redis
+  role: master
+
+#replicas: 1
```
Comment (cdrage) [inline examples/guestbook-go/redis-master.yml]: remove comment

File: examples/guestbook-go/guestbook.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+name: guestbook
+
+labels:
+  app: guestbook
+
+#replicas: 3
```
Comment (cdrage) [inline examples/guestbook-go/guestbook.yml]: remove commented out code

File: examples/guestbook-go/redis-slave.yml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+name: redis-slave
+
+labels:
+  app: redis
+  role: slave
+
+#replicas: 2
```
Comment (cdrage) [inline examples/guestbook-go/redis-slave.yml]: remove commented out line

File: examples/guestbook-official/README.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (cdrage) [inline examples/guestbook-official/README.md]: remove newlines

File: examples/guestbook-official/frontend.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  env:
+   - name: GET_HOSTS_FROM
+     value: dns
+
+services:
+- name: frontend
+  type: NodePort
+  portMappings:
+  - 80:80
+
```
Comment (cdrage) [inline examples/guestbook-official/frontend.yaml]: remove newline

File: examples/guestbook-official/redis-master.yaml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  resources:
+    requests:
+      cpu: 100m
+      memory: 100Mi
+
+services:
+- name: redis-master
+  portMappings:
+  - 6379:6379
+
```
Comment (cdrage) [inline examples/guestbook-official/redis-master.yaml]: remove newline

**Review by cdrage — CHANGES_REQUESTED** [review]

**Comment by cdrage** [comment]
Tested guestbook-official against Kubernetes :+1: 
Tested guestbook-go against Kubernetes :+1: 
Tested guestbook-go against OpenShift :+1: 

After you've fixed the extra newlines / commented out lines, feel free to merge in! :100:  Thanks @surajnarwade

File: examples/guestbook-official/frontend.yaml [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    requests:
+      cpu: 100m
+      memory: 100Mi
+  env:
+   - name: GET_HOSTS_FROM
+     value: dns
+
+services:
+- name: frontend
+  type: NodePort
```
Comment (cdrage) [inline examples/guestbook-official/frontend.yaml]: we should use loadbalancer rather than nodeport here

**Comment by cdrage** [comment]
@surajnarwade Just realized that NodePort is being used... this should use LoadBalancer: https://github.com/kubernetes/examples/blob/master/guestbook-go/guestbook-service.json

from the official example ^^

**Comment by cdrage** [comment]
another comment: they all need to be `.yaml` not `.yml`

**Comment by surajnarwade** [comment]
@cdrage , oh I forgot, I used `NodePort` for my convenience 
Updated with suggested changes

File: examples/guestbook-go/redis-slave.yaml [hunk 10]
```diff
@@ -0,0 +1,16 @@
+name: redis-slave
+
+labels:
+  app: redis
+  role: slave
+
+#replicas: 2
```
Comment (cdrage) [inline examples/guestbook-go/redis-slave.yaml]: remove commented out line

**Review by cdrage — CHANGES_REQUESTED** [review]
@surajnarwade one last change

**Review by kadel — CHANGES_REQUESTED** [review]
Containers on openshift are running, but redis replication is not working. And Guestbook is also not working (can't save messages)

**Comment by surajnarwade** [comment]
@kadel guestbook-go is not working for you ?

**Comment by kadel** [comment]
when I try to save message to guestbook i get following error in guestbook pod
```
[negroni] Started GET /rpush/guestbook/fff
--
  | [negroni] PANIC: MISCONF Redis is configured to save RDB snapshots, but is currently not able to persist on disk. Commands that may modify the data set are disabled. Please check Redis logs for details about the error.
  | goroutine 5 [running]:
  | github.com/codegangsta/negroni.(*Recovery).ServeHTTP.func1(0x7f6169388a90, 0xc82021ffc0, 0xc8200b6ea0)
  | /go/src/github.com/codegangsta/negroni/recovery.go:34 +0xe9
  | panic(0x76aee0, 0xc82030bff0)
  | /usr/local/go/src/runtime/panic.go:426 +0x4e9
  | main.HandleError(0x0, 0x0, 0x7f6169349608, 0xc82030bff0, 0x0, 0x0)
  | /go/src/github.com/GoogleCloudPlatform/kubernetes/examples/guestbook-go/_src/main.go:71 +0x59
  | main.ListPushHandler(0x7f6169388a90, 0xc82021ffc0, 0xc8202c0a80)
  | /go/src/github.com/GoogleCloudPlatform/kubernetes/examples/guestbook-go/_src/main.go:47 +0x192
  | net/http.HandlerFunc.ServeHTTP(0x8a8270, 0x7f6169388a90, 0xc82021ffc0,
```


all redist-slave pods are showing:
```

[7] 03 Jan 11:58:16.042 * Connecting to MASTER redis-master:6379
--
  | [7] 03 Jan 11:58:16.044 * MASTER <-> SLAVE sync started
  | [7] 03 Jan 11:58:16.044 * Non blocking connect for SYNC fired the event.
  | [7] 03 Jan 11:58:16.044 # Error reply to PING from master: '-MISCONF Redis is configured to save RDB snapshots, but is currently not able to persist on disk. Commands that may modify the data set are disabled. Please check Redis logs for details about the error.'
```

redis-master

```

[1] 03 Jan 11:51:54.310 * The server is now ready to accept connections on port 6379
--
  | [1] 03 Jan 11:52:29.677 * Slave 172.17.0.7:6379 asks for synchronization
  | [1] 03 Jan 11:52:29.677 * Full resync requested by slave 172.17.0.7:6379
  | [1] 03 Jan 11:52:29.677 * Starting BGSAVE for SYNC with target: disk
  | [1] 03 Jan 11:52:29.678 * Background saving started by pid 9
  | [9] 03 Jan 11:52:29.678 # Failed opening .rdb for saving: Permission denied
  | [1] 03 Jan 11:52:29.720 # Background saving error
  | [1] 03 Jan 11:52:29.720 # Connection with slave 172.17.0.7:6379 lost.
  | [1] 03 Jan 11:52:29.720 # SYNC failed. BGSAVE child returned an error


```

**Comment by cdrage** [comment]
I've been using this example fine, have you been waiting until redis-slave comes up (kubectl get po) and then testing the application on Guestbook / browser? I ran into the same issue of trying to save messages before it came up.

I'm using: https://github.com/kedgeproject/kedge/pull/554/files and it works.


@surajnarwade Can you check / compare between your example and https://github.com/kedgeproject/kedge/pull/554/files and see if it helps solve any issues you may be encountering?

File: examples/guestbook-go/guestbook.yaml [hunk 11]
```diff
@@ -0,0 +1,16 @@
+name: guestbook
+
+labels:
+  app: guestbook
+
+replicas: 3
```
Comment (cdrage) [inline examples/guestbook-go/guestbook.yaml]: remove replicas

File: examples/guestbook-go/redis-master.yaml [hunk 12]
```diff
@@ -0,0 +1,17 @@
+name: redis-master
+
+labels:
+  app: redis
+  role: master
+
+replicas: 1
```
Comment (cdrage) [inline examples/guestbook-go/redis-master.yaml]: remove replicas

File: examples/guestbook-go/redis-slave.yaml [hunk 13]
```diff
@@ -0,0 +1,16 @@
+name: redis-slave
+
+labels:
+  app: redis
+  role: slave
+
+replicas: 2
```
Comment (cdrage) [inline examples/guestbook-go/redis-slave.yaml]: remove replicas

File: examples/guestbook-official/frontend.yaml [hunk 14]
```diff
@@ -0,0 +1,24 @@
+name: frontend
+
+replicas: 3
```
Comment (cdrage) [inline examples/guestbook-official/frontend.yaml]: remove replicas (why is there 3?)

File: examples/guestbook-official/redis-master.yaml [hunk 15]
```diff
@@ -0,0 +1,21 @@
+name: redis-master
+
+replicas: 1
```
Comment (cdrage) [inline examples/guestbook-official/redis-master.yaml]: remove replicas

File: examples/guestbook-official/redis-slave.yaml [hunk 16]
```diff
@@ -0,0 +1,24 @@
+name: redis-slave
+
+replicas: 2
```
Comment (cdrage) [inline examples/guestbook-official/redis-slave.yaml]: remove replicas

**Review by cdrage — CHANGES_REQUESTED** [review]
When I commented I said to remove the line, not un-comment :) This should fix any issues that @kadel was having. This modification works for me (see: https://github.com/kedgeproject/kedge/pull/554/files )

**Comment by surajnarwade** [comment]
@kadel can you try once again ?

**Comment by kadel** [comment]
guestbook-go is still not working on OpenShift. (works if root is enabled, but that is not what we want)

redis-master:
```
[8] 05 Jan 10:16:19.629 # Failed opening .rdb for saving: Permission denied
```

Haven't look at it closer, but I suspect that there are some problems with file or directory permissions. Maybe adding PVC where redis data will be saved might help?

**Comment by surajnarwade** [comment]
I tried,

```
$ kedge apply -f examples/guestbook-go/
W0116 14:31:22.724313   24483 factory_object_mapping.go:423] Failed to download OpenAPI (unknown), falling back to swagger
service "guestbook" created
W0116 14:31:22.943018   24492 factory_object_mapping.go:423] Failed to download OpenAPI (unknown), falling back to swagger
deployment "guestbook" created
W0116 14:31:23.152870   24503 factory_object_mapping.go:423] Failed to download OpenAPI (unknown), falling back to swagger
service "redis-master" created
W0116 14:31:23.365897   24514 factory_object_mapping.go:423] Failed to download OpenAPI (unknown), falling back to swagger
deployment "redis-master" created
W0116 14:31:23.559956   24524 factory_object_mapping.go:423] Failed to download OpenAPI (unknown), falling back to swagger
service "redis-slave" created
W0116 14:31:23.757599   24533 factory_object_mapping.go:423] Failed to download OpenAPI (unknown), falling back to swagger
deployment "redis-slave" created
```

I got,

```
$ oc get pods
NAME                            READY     STATUS    RESTARTS   AGE
guestbook-871169334-50bbl       1/1       Running   0          19m
redis-master-2082469664-f1h3v   1/1       Running   0          19m
redis-slave-3332111353-tkhpc    1/1       Running   0          19m
```

No addons is enabled,

```
$ minishift addons list
- admin-user     : disabled   P(0)
- anyuid         : disabled   P(0)
- registry-route : disabled   P(0)
- xpaas          : disabled   P(0)
```

**Comment by surajnarwade** [comment]
@kadel , examples is working fine on my system :D

**Comment by kadel** [comment]
It is nice that pods are running, but that doesn't mean that application is working!

Have you actually tried to access the guestbook web and save some message?
Have you looked at logs and verified that redis is working?

**Comment by surajnarwade** [comment]
@kadel you are right, only guestbook is running but redis is failing

**Comment by surajnarwade** [comment]
what should we do about this ?

**Comment by cdrage** [comment]
@surajnarwade Does it currently work? I'd say another review and we can try and get this in?

**Pull request closed without merging**